### PR TITLE
Fix qualified modifier lookup.

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/10303

No changelog entry since this is only present in breaking.

TODO:
 - [x] document how cross-contract access works and whether or not modifiers from libraries are accessible

### CODE REVIEW

File: docs/grammar/Solidity.g4 [hunk 1]
```diff
@@ -329,7 +329,6 @@ expression:
 		identifier
 		| literal
 		| elementaryTypeName[false]
-		| userDefinedTypeName
```
Comment (hrkrshnn) [inline docs/grammar/Solidity.g4:332]: Why was this removed?

**Comment by cameel** [comment]
I just noticed that it's possible to define modifiers inside libraries and interfaces and the compiler does not complain. For libraries it's probably fine but in case of interfaces it seems to be a bug.

A side-effect of the change in the grammar here is that it's now possible to actually use these modifiers.

```solidity
library L {
    modifier m {
        _;
    }
}

contract C {
    function f() public L.m { }
}
```

On 0.7.4:
```
Error: Expected '{' but got '.'
 --> /tmp/test.sol:8:26:
  |
8 |     function f() public L.m { }
  |                          ^
```

After this PR there are no errors.

Do we want this? If so, I think we should mention it in the docs and in the changelog and also add a test covering this usage.

File: test/libsolidity/semanticTests/modifiers/access_through_contract_name.sol [hunk 2]
```diff
@@ -0,0 +1,22 @@
+contract A {
+    uint public x = 7;
+    modifier m virtual { x = 2; _; }
+}
+contract C is A {
+    modifier m override { x = 1; _; }
+
+    function f() public A.m returns (uint) {
```
Comment (cameel) [inline test/libsolidity/semanticTests/modifiers/access_through_contract_name.sol:8]: How about a syntax test with a few more dots?

`M.sol`
```solidity
import "M.sol" as M;

contract C {
    modifier m {
        _;
    }

    function f() public M.M.M.C.m {}
}
```

File: docs/grammar/Solidity.g4 (hunk 1 above)
Comment (chriseth) [reply docs/grammar/Solidity.g4:332]: Because it is not parsed like that, the expression parser uses `MemberAccess` to parse `.`.

**Comment by chriseth** [comment]
@cameel I think we should not allow cross-contract modifier access, except maybe for internal library modifiers...

**Comment by cameel** [comment]
> except maybe for internal library modifiers...

Oh, looks like these do work on 0.7.4.

**Comment by cameel** [comment]
The code seems fine now but your error code is not unique.

**Comment by chriseth** [comment]
TODO: document changes (breaking: modifiers of libraries are not accessible anymore)

**Comment by cameel** [comment]
> TODO: document changes (breaking: modifiers of libraries are not accessible anymore)

Is it still breaking if qualified names didn't work before anyway due to a bug?
The internal unqualified names worked and still do work so they are accessible in some situations.

I think we should just mention where you can define/use modifiers on the documentation page about them.

File: test/libsolidity/semanticTests/modifiers/access_through_module_name.sol [hunk 3]
```diff
@@ -0,0 +1,25 @@
+==== Source: a ====
+import "a" as M;
+contract C {
+    uint public x;
+    modifier m { x = 1; _; }
+
+    function f() public M.M.C.m returns (uint t, uint r) {
```
Comment (bshastry) [inline test/libsolidity/semanticTests/modifiers/access_through_module_name.sol:7]: Shouldn't the modifier identifier path be `M.C.m`?

File: test/libsolidity/syntaxTests/modifiers/access_in_library.sol [hunk 4]
```diff
@@ -0,0 +1,9 @@
+library L {
+    modifier m() { _; }
+}
+contract C {
+    function f() L.m public {
+    }
+}
+// ----
+// TypeError 9428: (68-71): Can only use modifiers defined in the current contract or in base contracts.
```
Comment (bshastry) [inline test/libsolidity/syntaxTests/modifiers/access_in_library.sol:9]: Does `using` declaration permit access to modifiers defined in libraries?

```
library L {
    modifier m() { _; }
}
contract C {
    using L for *;
    function f() L.m public {
    }
}
```

File: test/libsolidity/semanticTests/modifiers/access_through_module_name.sol (hunk 3 above)
Comment (chriseth) [reply test/libsolidity/semanticTests/modifiers/access_through_module_name.sol:7]: `M` is visible inside itself, so it was added just for the fun of it

File: test/libsolidity/syntaxTests/modifiers/access_in_library.sol (hunk 4 above)
Comment (chriseth) [reply test/libsolidity/syntaxTests/modifiers/access_in_library.sol:9]: i don't think so - using does not work like that, but I can add such a test.

**Comment by chriseth** [comment]
@cameel you are right ,this just was not possible before!

File: docs/contracts/function-modifiers.rst [hunk 5]
```diff
@@ -15,6 +15,11 @@ inheritable properties of contracts and may be overridden by derived contracts,
 if they are marked ``virtual``. For details, please see
 :ref:`Modifier Overriding <modifier-overriding>`.
 
+If you want to access the modifier defined in the current or a base contract ``C``,
+you can use ``C.m`` to reference the modfier defined in ``C`` without
```
Comment (cameel) [inline docs/contracts/function-modifiers.rst]: A typo:
```suggestion
you can use ``C.m`` to reference the modifier defined in ``C`` without
```

Comment (cameel) [reply docs/contracts/function-modifiers.rst]: Also, what do you think about adding more detail here?
```suggestion
If you want to access a modifier `m` defined in a contract ``C``, you can use ``C.m`` to
reference it without virtual lookup. It's only possible to use modifiers defined in the current
contract or its base contracts. Modifiers can also be defined in libraries but their use is
limited to functions of the same library.
```

**Review by cameel — COMMENTED** [review]
Just a small tweak in the docs.

**Comment by chriseth** [comment]
Updated.

**Review by cameel — APPROVED** [review]
